docs/rfc/rfc-105-event-observer.md

```diff
@@ -0,0 +1,240 @@
+their own concurrency.
+
+### Fanout Observers
+
+How then does one simulate the same sort of pub/sub functionality that the
+`EventBus` provides using this approach? Where this sort of behaviour is
+absolutely necessary, it is trivial to build a "fanout" observer that implements
+this interface, e.g.:
+
+```golang
+type StateFanoutObserver struct {
+    observers []StateObserver
+}
+
+func NewStateFanoutObserver(observers ...StateObserver) *StateFanoutObserver {
+    return &StateFanoutObserver{
+        observers: observers,
+    }
+}
+
+func (o *StateFanoutObserver) ConsensusStateNewRoundStep(ev EventDataRoundStep) {
+    for _, obs := range o.observers {
+        obs.ConsensusStateNewRoundStep(ev)
+    }
+}
+
+// ...
+```
+
+### Testing
+
+Many tests in the CometBFT codebase rely heavily on subscribing to specific
+events directly via the event bus. This can easily be accomplished using the
+approach described in the [Fanout Observers](#fanout-observers) section:
+
+```golang
+state.SetObserver(
+    NewStateFanoutObserver(
+        // An observer specifically for use during testing.
+        newTestStateObserver(),
+        // ... other observers here that would be used in production
+    ),
+)
+```
+
+One could easily also define an ergonomic observer type that would allow inline
+definition and overriding of only specific event handlers:
+
+```golang
+type testObserver struct {
+    newRoundStep   func(EventDataRoundState)
+    timeoutPropose func(EventDataRoundState)
+    // ...
+}
+
+func (o *testObserver) ConsensusStateNewRoundStep(ev EventDataRoundState) {
+    if o.newRoundStep != nil {
+        o.newRoundStep(ev)
+    }
+}
+
+// ...
+
+func TestCustomObserver(t *testing.T) {
+    testObs := &testObserver{}
+    testObs.newRoundStep = func(ev EventDataRoundState) {
+        // Custom code here called upon new round step
+    }
+    // ...
+}
+```
+
+### Event Subscription
+
+The current WebSocket-based event subscription mechanism is envisaged to go away
+at some point in future, and there is no other mechanism by which external
+observers can subscribe to events.
+
+[ADR 101], however, provides a more general alternative through which
+integrators can gain access to event data from outside of the node. Once ADR 101
+has been implemented, the whole WebSocket-based interface could be removed.
+
+### Pros and Cons
+
+The benefits of the proposed approach include:
+
+- Greater compile-time correctness guarantees
+- Code becomes easier to reason about, since one can easily follow the call
+  chain for certain events using one's IDE instead of needing to search the
+  codebase for subscriptions to certain events
+- Easier to test
+- Does away with needing to access internal/private `eventBus` variables within
+  reactors/state from tests ([example][test-eventbus-access])
+- Splits event generation and handling out into a per-package responsibility,
+  more cleanly separating and modularizing the codebase
+
```

Comment:
A "conclusion" bullet here would be the KISS principle. When a simpler solution fits your needs, if you go for the complex one you're over-engineering .